prerendering-state.md

```diff
@@ -0,0 +1,150 @@
+
+### Web-compatibility
+
+Related to the above point, a common pattern on the web is to assume only two visibility states. For example, code such
+as:
+
+```
+onload = () => {
+  if (document.visibilityState == 'visible')
+    // Load large images
+  else
+    // Defer loading large images
+}
+```
+
+An analysis [found several
+examples](https://docs.google.com/document/d/1Xzw0k8DgltI2ohapuDKmjRZLv7NVrRFGusW8IBtiCT0/edit#heading=h.rkorueof7xev)
+of such behavior.
+
+(re-)Introducing a `prerender` state could break such pages. In particular, portals (or some other, future
+prerendering-while-visible context) would show static or missing content. A hidden prerender may not have immediate
+user-visible effects, but may use more resources than necessary.
+
+## document.prerendering
+
+We propose adding a new state variable:
+
+```
+partial interface Document {
+    readonly attribute boolean    prerendering;
+    attribute EventHandler        onprerenderingchange;
+};
+```
+
+`document.prerendering` returns true if the document is currently hosted in a [prerendering browsing
+context](browsing-context.md). The `prerenderingchange` event is fired whenever this value changes.
+
+A state separate from visibility ensures existing visibility-based states are correctly accounted for. It also forces
+authors to consider their page's behavior in light of prerendering restrictions and non-interactivity and the fact that
+a user may not have initiated the page load in the first place.
+
+### Relationship to visibilityState
+
+`prerendering` and `visibilityState` are entirely independent and all combinations of states are possible:
+
+| State          | `prerendering` | `visibilityState` |
+| -------------- | -------------- | ----------------- |
+| Foreground Tab | false          | 'visible'         |
+| Background Tab | false          | 'hidden'          |
+| Portal         | true           | 'visible'         |
+| Prerender      | true           | 'hidden'          |
+
+#### VisibilityState as a termination signal
+
+Authors have been [encouraged](https://www.igvita.com/2015/11/20/dont-lose-user-and-app-state-use-page-visibility/) to
+switch from `unload` and `beforeunload` events to `visibilityState=='hidden'` as a signal that their page may be
+terminated. See also [PageVisibility#59](https://github.com/w3c/page-visibility/issues/59) for related discussion.
+
+Entering a `prerendering` state, via being put into a portal, is likely a point where such a termination signal should
+be fired; it is analogous to navigating to a new page. However, this means pages would now also have to listen for a
+switch to `prerendering == true`.
+
+This [background
+doc](https://docs.google.com/document/d/1Xzw0k8DgltI2ohapuDKmjRZLv7NVrRFGusW8IBtiCT0/edit#heading=h.acmnp6zdmcik) deals
+extensively with this issue. An ideal solution would introduce a new event for this use case, rather than implicitly
+tying it to any particular state.
+
+### Relationship to loading-modes
```

Comment:
Let's delete this section and roll it into the TODO you inserted above, which I commented on.

Reply:
Done